bofire/strategies/doe/objective.py

```diff
@@ -0,0 +1,459 @@
+            The jacobian of -1 * min(eigvals(X.T @ X + delta)) as numpy array
+        """
+        # get model matrix X
+        X = self._convert_input_to_model_tensor(x, requires_grad=True)
+
+        # first part of jacobian
+        torch.min(
+            torch.linalg.eigvalsh(X.T @ X + self.delta * torch.eye(self.n_model_terms))
+        ).backward()
+        J1 = -1 * X.grad.detach().numpy()  # type: ignore
+        J1 = np.repeat(J1, self.n_vars, axis=0).reshape(
+            self.n_experiments, self.n_vars, self.n_model_terms
+        )
+
+        # second part of jacobian
+        J2 = self._model_jacobian_t(x)
+
+        # combine both parts
+        J = J1 * J2
+        J = np.sum(J, axis=-1)
+
+        return J.flatten()
+
+
+class KOptimality(Objective):
+    """A class implementing the evaluation of the condition number of (X.T @ X + delta)
+    and its jacobian w.r.t. the inputs. The jacobian evaluation is done analogously to
+    DOptimality with the first part of the jacobian being the jacobian of condition number
+    of (X.T @ X + delta) instead of logdet(X.T@X + delta).
+    """
+
+    def evaluate(self, x: np.ndarray) -> float:
+        """Computes condition number of (X.T@X + delta).
+        Where X is the model matrix corresponding to x.
+
+        Args:
+            x (np.ndarray): values of design variables a 1d array.
+
+        Returns:
+            cond(X.T @ X + delta)
+        """
+        X = self._convert_input_to_model_tensor(x, requires_grad=False)
+        return float(
+            torch.linalg.cond(
+                X.detach().T @ X.detach() + self.delta * torch.eye(self.n_model_terms)
+            )
+        )
+
+    def evaluate_jacobian(self, x: np.ndarray) -> np.ndarray:
+        """Computes the jacobian of the condition number of (X.T @ X + delta).
+        Where X is the model matrix corresponding to x.
+
+        Args:
+            x (np.ndarray): values of design variables a 1d array.
+
+        Returns:
+            The jacobian of cond(X.T @ X + delta) as numpy array
+        """
+        # get model matrix X
+        X = self._convert_input_to_model_tensor(x, requires_grad=True)
+
+        # first part of jacobian
+        torch.linalg.cond(
+            X.T @ X + self.delta * torch.eye(self.n_model_terms)
+        ).backward()
+        J1 = X.grad.detach().numpy()  # type: ignore
+        J1 = np.repeat(J1, self.n_vars, axis=0).reshape(
+            self.n_experiments, self.n_vars, self.n_model_terms
+        )
+
+        # second part of jacobian
+        J2 = self._model_jacobian_t(x)
+
+        # combine both parts
+        J = J1 * J2
+        J = np.sum(J, axis=-1)
+
+        return J.flatten()
+
+
+# TODO: test
+class SpaceFilling(Objective):
+
+    # TODO: how many values should be taken into account. 1 is not enough. Right now it is n_experiments, this seems to work.
+    def evaluate(self, x: np.ndarray) -> float:
+        X = self._convert_input_to_tensor(x, requires_grad=False)
```

Comment:
`_convert_input_to_tensor` computes the model matrix, space filling should be model independent, only consider (x1, x2) and not e.g. (x1, x2, x1*x2, x1**2, x2**2) for the distance calculation.

Reply:
No, `_convert_input_to_tensor()` is slightly different.
```
    def _convert_input_to_tensor(
        self, x: np.ndarray, requires_grad: bool = True
    ) -> Tensor:
        X = pd.DataFrame(
            x.reshape(len(x.flatten()) // self.n_vars, self.n_vars), columns=self.vars
        )
        return torch.tensor(X.values, requires_grad=requires_grad, **tkwargs)
```
It misses the `get_model_matrix()` call that `_convert_input_to_model_tensor()` has. But I have not forgotten that you wanted a separate class for the `SpaceFilling`! This was just a quick experiment if this objective works, will convert the PR to a draft for now.

Reply:
True. Didn't see that `_convert_to_model_tensor` is redefined in `SpaceFilling`.